From 8a58bff3e4795942ffde5380b7d658e9eb61df3e Mon Sep 17 00:00:00 2001
From: Bin Jiang <bin.jiang@windriver.com>
Date: Tue, 3 Jun 2014 18:29:09 +0800
Subject: [PATCH 1/3] ARM: iProc: add irqflags with IRQF_SHARED value.

If startup/shutdown serial repeatedly, there will be following call trace:
irq 123: nobody cared (try booting with the "irqpoll" option)
CPU: 0 PID: 42 Comm: irq/123-GPIOA Not tainted 3.10.38-ltsi-rt34-WR6.0.0.0_preempt-rt #20
[<c0017f78>] (unwind_backtrace+0x0/0x100) from [<c00130f8>] (show_stack+0x20/0x24)
[<c00130f8>] (show_stack+0x20/0x24) from [<c05cc834>] (dump_stack+0x24/0x28)
[<c05cc834>] (dump_stack+0x24/0x28) from [<c009d228>] (__report_bad_irq+0x38/0xd0)
[<c009d228>] (__report_bad_irq+0x38/0xd0) from [<c009d758>] (note_interrupt+0x18c/0x244)
[<c009d758>] (note_interrupt+0x18c/0x244) from [<c009bf4c>] (irq_thread+0x1b8/0x1c8)
[<c009bf4c>] (irq_thread+0x1b8/0x1c8) from [<c0049460>] (kthread+0xb4/0xb8)
[<c0049460>] (kthread+0xb4/0xb8) from [<c000eb58>] (ret_from_fork+0x14/0x20)

When startup a serial, followings will run:
  serial8250_startup
	|
	+ disable_irq_nosync() (*1)
	|
	+ wait_for_xmitr() (*2)
	|
	+ enable_irq() (*3)
	|
	+ request_irq()

(*1) and (*3) run on condition that the irq is shared.
At (*2), wait for transmitter & holding register to empty and a THRE
interrupt will trigger. For shared irq, (*2) must run in irqs disabled
context, otherwise no any irq handlers can handle the THRE interrupt and
there will be an irq storm and above call trace.

Since the GPIO and UART share an irq, set UART' irqflags as IRQF_SHARED.

Signed-off-by: Bin Jiang <bin.jiang@windriver.com>
---
 arch/arm/mach-iproc/common.c |    1 +
 1 files changed, 1 insertions(+), 0 deletions(-)

diff --git a/arch/arm/mach-iproc/common.c b/arch/arm/mach-iproc/common.c
index a3291bf..b120a70 100644
--- a/arch/arm/mach-iproc/common.c
+++ b/arch/arm/mach-iproc/common.c
@@ -69,6 +69,7 @@
 	.iotype     = UPIO_MEM,						\
 	.type       = PORT_16550A,					\
 	.flags      = UPF_FIXED_TYPE | UPF_SHARE_IRQ,			\
+	.irqflags   = IRQF_SHARED,					\
 	.private_data = (void __iomem *)((IPROC_##name##_VA) + 0x00),	\
 }
 
-- 
1.7.5.4

